java-checks/src/main/java/org/sonar/java/checks/NullShouldNotBeUsedWithOptionalCheck.java

```diff
@@ -0,0 +1,132 @@
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.AnnotationTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.ConditionalExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.LambdaExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.ModifiersTree;
+import org.sonar.plugins.java.api.tree.ReturnStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.Tree.Kind; 
+
+@Rule(key = "S2789")
+public class NullShouldNotBeUsedWithOptionalCheck extends BaseTreeVisitor implements JavaFileScanner {
+  
+  private static final String NULLABLE = "javax.annotation.Nullable";
+  
+  private static final String OPTIONAL = "java.util.Optional";
+
+  private JavaFileScannerContext context;
+
+  @Override
+  public void scanFile(final JavaFileScannerContext context) {
+    this.context = context;
+    scan(context.getTree());
+  }
+
+  @Override
+  public void visitMethod(MethodTree method) {
+    if (!method.is(Tree.Kind.CONSTRUCTOR) && returnsOptional(method)) {
+
+      // check that the method is not annotated with @Nullable
+      ModifiersTree modifiers = method.modifiers();
+      for (AnnotationTree annotation : modifiers.annotations()) {
+        Type type = annotation.annotationType().symbolType();
+        if (type.is(NULLABLE)) {
+          context.reportIssue(this, annotation, "Methods with an \"Optional\" return type should not be \"@Nullable\".");
+        }
+      }
+
+      // check that the method does not return "null"
+      method.accept(new ReturnNullVisitor());
+    }
+
+    super.visitMethod(method);
+  }
+
+  @Override
+  public void visitBinaryExpression(BinaryExpressionTree binaryExpression) {
+    String message = "Remove this null-check of an \"Optional\".";
+    if (binaryExpression.is(Tree.Kind.EQUAL_TO, Tree.Kind.NOT_EQUAL_TO)) {
+      ExpressionTree left = binaryExpression.leftOperand();
+      ExpressionTree right = binaryExpression.rightOperand();
+      if ((isOptional(left) && isNull(right)) || (isNull(left) && isOptional(right))) {
+        context.reportIssue(this, binaryExpression, message);
+      }
+    }
+
+    super.visitBinaryExpression(binaryExpression);
+  }
+
+  private class ReturnNullVisitor extends BaseTreeVisitor {
+
+    @Override
+    public void visitReturnStatement(ReturnStatementTree returnStatement) {
+      ExpressionTree expression = returnStatement.expression();
+      if (!checkNull(expression) && expression.is(Kind.CONDITIONAL_EXPRESSION)) {
+        ConditionalExpressionTree conditionalExpression = (ConditionalExpressionTree)expression; 
```

Comment:
You should extract this part into the `visitConditionalExpression` method, and start the method by checking if the type of the expression is `Optional`.